third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1,35 +1,232 @@
+// (2) outer reduction is same to cases where hidden_size > 1024 except the
+// second step where in this case, the reduction domain is parallelized by TIDx
+// and the iteration domain is parallelized by TIDy and BIDy. This switch
+// between TIDx and TIDy is because (a) we can do warp reduction with TIDx and
+// (b) TIDx*BIDy is usually much larger than hidden_size, e.g. 128*216 = 1024*27
+// this means without switch only 1/27 of the threads is used.
+std::shared_ptr<ReductionParams> innerOuterPersistentHeuristic(
+    const int64_t outer_dim_numel,
+    const int64_t inner_dim_numel,
+    const int64_t n_tensor_inputs,
+    const int64_t max_input_dtype_size,
+    const int64_t max_persistent_buffer_size,
+    const size_t vectorize_factor) {
+  auto rparams = std::make_shared<ReductionParams>();
+  rparams->persistent_kernel = true;
+  rparams->fastest_dim = true;
+  rparams->combined_inner_outer = true;
+
+  const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  const int64_t device_multiprocessor_count =
+      (int64_t)dev_prop->multiProcessorCount;
+
+  // step-1, reduction dim of inner reduction, vect * batch * threads_per_block
+  // >= inner_dim_numel vect represents unroll_factor_inner_reduction
+  int64_t vect = vectorize_factor;
+  int64_t after_vect = ceilDiv(inner_dim_numel, vect);
+
+  // split after_vect between threads_per_block and batch
+  // tuned for layer_norm backward on A100
+  // large batch size leads small block size and low intra-threads
+  // communications with more work per thread small batch size leads large block
+  // size and high intra-threads communications with less work per thread
+  // batch represents batches_per_block_inner_reduction
+  int64_t threads_per_block = 32;
+  int64_t batch = ceilDiv(after_vect, threads_per_block);
+  int64_t batch_max = 20;
+  int64_t batch_min;
+  if (inner_dim_numel >= 3072) {
+    // allow small batch for small iter numbers, e.g. shape= 2048 x 3072
+    if (outer_dim_numel <= 2048 && inner_dim_numel == 3072) {
+      batch_min = 3;
+    } else {
+      batch_min = 4;
+    }
+  } else if (inner_dim_numel >= 2048) {
+    batch_min = 2;
+  } else {
+    batch_min = 1;
+  }
+  // split workload between batch and threads_per_block
+  // decrease batch and increase threads_per_block while current bath is larger
+  // batch_min and current threads_per_block is larger than
+  // threads_per_block_max
+  int64_t threads_per_block_max = inner_dim_numel >= 20480 ? 512 : 256;
+  while (batch > batch_min && threads_per_block < threads_per_block_max) {
+    if (batch % 2 == 0 && batch / 2 >= batch_min) {
+      batch /= 2;
+      threads_per_block *= 2;
+    } else if (batch % 3 == 0 && batch / 3 >= batch_min) {
+      batch /= 3;
+      threads_per_block *= 3;
+    } else if (batch % 5 == 0 && batch / 5 >= batch_min) {
+      batch /= 5;
+      threads_per_block *= 5;
+    } else if (batch > batch_max) {
```

Comment:
It seems in this case `threads_per_block` could exceed `threads_per_block_max`. Is it fine?

Reply:
Actually, should this be placed inside this while loop? Looks like this makes sure `batch` never exceeds `batch_max`. Why not just do it once before this while loop?